op-node/rollup/status/l1_tracker_test.go

```diff
@@ -0,0 +1,108 @@
+package status
+
+import (
+	"context"
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/testutils"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+func mockL1BlockRef(num uint64) eth.L1BlockRef {
+	return eth.L1BlockRef{Number: num, Hash: common.Hash{byte(num)}, ParentHash: common.Hash{byte(num - 1)}}
+}
+
+func newL1HeadEvent(l1Tracker *L1Tracker, head eth.L1BlockRef) {
+	l1Tracker.OnEvent(L1UnsafeEvent{
+		L1Unsafe: head,
+	})
+}
+
+func TestCachingReorgs(t *testing.T) {
+	ctx := context.Background()
+	l1Fetcher := &testutils.MockL1Source{}
+	l1Tracker := NewL1Tracker(l1Fetcher)
+
+	// no blocks added to cache yet
+	l1Head := mockL1BlockRef(99)
+	l1Fetcher.ExpectL1BlockRefByNumber(99, l1Head, nil)
+	ret, err := l1Tracker.L1BlockRefByNumber(ctx, 99)
+	require.NoError(t, err)
+	require.Equal(t, l1Head, ret)
+	l1Fetcher.AssertExpectations(t)
+
+	// from cache
+	l1Head = mockL1BlockRef(100)
+	newL1HeadEvent(l1Tracker, l1Head)
+	ret, err = l1Tracker.L1BlockRefByNumber(ctx, 100)
+	require.NoError(t, err)
+	require.Equal(t, l1Head, ret)
+
+	// from cache
+	l1Head = mockL1BlockRef(101)
+	newL1HeadEvent(l1Tracker, l1Head)
+	ret, err = l1Tracker.L1BlockRefByNumber(ctx, 101)
+	require.NoError(t, err)
+	require.Equal(t, l1Head, ret)
+
+	// from cache
+	l1Head = mockL1BlockRef(102)
+	newL1HeadEvent(l1Tracker, l1Head)
+	ret, err = l1Tracker.L1BlockRefByNumber(ctx, 102)
+	require.NoError(t, err)
+	require.Equal(t, l1Head, ret)
+
+	// trigger a reorg of block 101, invalidating the following cache elements
+	l1Head = mockL1BlockRef(101)
```

Comment:
We should add tests for the other types of reorg possibilities as well where the common ancestor isn't the parent of the new head and where the chain gets shorter, longer or stays the same length during the reorg.